packages/next/next-server/server/next-server.ts

```diff
@@ -485,15 +485,18 @@ export default class Server {
           type: route.type,
           name: `${route.type} ${route.source} header route`,
           fn: async (_req, res, params, _parsedUrl) => {
+            const withParams = Object.keys(params).length > 0
             for (const header of (route as Header).headers) {
               let { key, value } = header
-              if (key.includes(':')) {
-                // see `prepareDestination` util for explanation for
-                // `validate: false` being used
-                key = compilePathToRegex(key, { validate: false })(params)
-              }
-              if (value.includes(':')) {
-                value = compilePathToRegex(value, { validate: false })(params)
+              if (withParams) {
+                if (key.includes(':')) {
+                  // see `prepareDestination` util for explanation for
+                  // `validate: false` being used
+                  key = compilePathToRegex(key, { validate: false })(params)
+                }
+                if (value.includes(':')) {
+                  value = compilePathToRegex(value, { validate: false })(params)
+                }
```

Comment:
I think we still need to handle the case of there being params and one of the values being a URL which tries to use the params e.g.
```js
{
  source: '/:path*',
  headers: [
    {
      key: 'access-control-allow-origin',
      value: 'https://example.com/:path*'
    }
  ]
}
```

If we detect that the value is a URL we can re-use the `prepareDestination` util used for the redirects/rewrites

Reply:
Escape URLs for path-to-regexp.

Reply:
`prepareDestination` is hard to use as it is because the spaces are escaped.